components/raftstore/src/store/peer.rs

```diff
@@ -3828,6 +3941,84 @@ where
         Ok(prs)
     }
 
+    pub fn maybe_witness_transfer_leader<T>(&mut self, ctx: &PollContext<EK, ER, T>) {
+        if !(self.is_witness() && self.is_leader()) {
+            return;
+        }
+        fail_point!("witness_transfer_leader", |_| {});
+        info!(
+            "witness try to transfer leader";
+            "region_id" => self.region_id,
+            "peer_id" => self.peer.get_id(),
+        );
+
+        if self.raft_group.raft.lead_transferee.is_some() {
+            info!("witness already transferring leader");
+            // already being in transfer leader
+            return;
+        }
+
+        let prs = self.raft_group.raft.prs();
```

Comment:
It's not responsive enough for PD to transfer leader for witness, as PD checks the regions one by one periodically and has a transfer-leader scheduling limit which avoids the operator from being added in time.

Reply:
> not responsive enough

Any tests to show the difference? Usually the limit is not the bottleneck but the time of transfer leader's cost is.

Reply:
No test. It's decided by the PD's mechanism. Polling regions a round needs hundreds of milliseconds or even more. It never gives promise that the operator is scheduled at once. And what if there is an existing operator on that region?

Reply:
> And what if there is an existing operator on that region?

It doesn't matter. Just like adding an evict leader on a store, all leaders will be evicted right away.

Reply:
But it's not fast enough. What do you think @rleungx

Reply:
What the speed requirement? I prefer to let PD to handle all the leader transfer because PD is supposed to be the component where scheduling decision is made. Every components try to transfer leader will make the system hard to determine and break configured constraints.

Reply:
I agree that all scheduling decisions should be made by PD. But currently, when the cluster is large, it usually takes several minutes to traverse all regions. Is that accepted?

Reply:
several minutes is really unacceptable

Reply:
@BusyJay I agree the target TiKV chooses may break configured constraints. But placement rule's following operator would schedule it to the right place.